uaclient/api/u/pro/attach/auto/full_auto_attach/v1.py

```diff
@@ -0,0 +1,126 @@
+            instance=cloud
+        )
+    except exceptions.ContractAPIError as e:
+        if e.code and 400 <= e.code < 500:
+            raise exceptions.NonAutoAttachImageError(
+                messages.UNSUPPORTED_AUTO_ATTACH
+            )
+        raise e
+
+    token = tokenResponse["contractToken"]
+    return token
+
+
+def _is_any_beta(cfg: UAConfig, ents: List[str]) -> Tuple[bool, str]:
+    for name in ents:
+        try:
+            ent_cls = entitlements.entitlement_factory(cfg, name)
+            if ent_cls.is_beta:
+                return True, name
+        except exceptions.EntitlementNotFoundError:
+            continue
+    return False, ""
+
+
+def full_auto_attach(options: FullAutoAttachOptions) -> FullAutoAttachResult:
+    return _full_auto_attach(options, UAConfig(root_mode=True))
+
+
+def _full_auto_attach(options: FullAutoAttachOptions, cfg: UAConfig):
+    instance = actions.get_cloud_instance(cfg)
+    allow_enable = options.enable is None and options.enable_beta is None
+    token = _get_auto_attach_token(cfg, instance)  # type: ignore
+    actions.attach_with_token(cfg, token=token, allow_enable=allow_enable)
+    if allow_enable:
+        return FullAutoAttachResult()
+
+    services = set()
+    if options.enable:
+        is_beta_found, service = _is_any_beta(cfg, options.enable)
+        if is_beta_found:
+            raise exceptions.BetaServiceError(
+                msg="beta service found in the enable list",
+                msg_code="beta-service-found",
+                additional_info={"beta_service": service},
+            )
+        services.update(options.enable)
+    if options.enable_beta:
+        services.update(options.enable_beta)
+
+    service_list = list(set(services))
+    found, not_found = entitlements.get_valid_entitlement_names(
+        service_list, cfg
+    )
+
+    for name in found:
+        ent_ret, reason = actions.enable_entitlement_by_name(
+            cfg, name, assume_yes=True, allow_beta=True
+        )
+        if not ent_ret:
+            if (
+                reason is not None
+                and isinstance(reason, CanEnableFailure)
+                and reason.message is not None
+            ):
+                raise exceptions.EntitlementNotEnabledError(
+                    msg=reason.message.msg,
+                    msg_code=reason.message.name,
+                    additional_info={"service": name},
+                )
+            else:
+                raise exceptions.EntitlementNotEnabledError(
+                    msg="Failed to enable service: {}".format(name),
+                    msg_code="entitlement-not-enabled",
+                    additional_info={"service": name},
+                )
+
+    if not_found:
+        msg = entitlements.create_enable_entitlements_not_found_message(
+            not_found, cfg=cfg, allow_beta=True
+        )
+        raise exceptions.EntitlementNotFoundError(msg.msg, not_found)
```

Comment:
An idea to discuss with the team. Some of these checks we can do up front. even before auto-attaching. Should we do a full verification of the requested services and raise an error before doing anything? @CalvoM @lucasmoura @renanrodrigo

Reply:
Or alternatively, should we move all checks as late as possible? Like we could move the beta service error to the end

Reply:
@CalvoM this validation is still being treated differently from the beta and incompatible checks. Should we move it up before any attaching happens?

Reply:
Additionally, we should have a better way to check for incompatible services in the list as @renanrodrigo mentioned, like `["fips", "livepatch"]`. But maybe it should be a follow up PR

Reply:
@lucasmoura I added the `_is_compatible_services_present` function, you can check it out.